### PR TITLE
feat: port to macfuse

### PR DESCRIPTION
**Pull request by killagu**

### CODE REVIEW

File: src/transport/macfuse/session.rs [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+            return Err(SessionFailure(String::from("failed to mount")))
+        }
+        let file: File = File::from_raw_fd(fd);
+        return Ok(file);
+    }
+}
+
+fn create_disk(mountpoint: &Path, dasession: DASessionRef) -> DADiskRef {
+    unsafe {
+        let mut mountpoint_argv = String::from(mountpoint.to_str().unwrap());
```
Comment (xujihui1985) [inline src/transport/macfuse/session.rs:272]: is it better to propagate the error instead of unwrap?

File: src/transport/macfuse/session.rs [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+        let mut mountpoint_argv = CString::new(mountpoint_argv).unwrap();
+        let mut mountpoint_argv = mountpoint_argv.as_ptr();
+        let url_str = CFStringCreateWithBytes(
+            std::ptr::null(),
+            std::mem::transmute(mountpoint_argv),
```
Comment (xujihui1985) [inline src/transport/macfuse/session.rs:277]: or just mountpoint_argv.as_ptr() as *const u8  instead?

File: src/transport/macfuse/session.rs [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+            mountpoint_argv.into_raw(),
+        ];
+        // let mut argv = ManuallyDrop::new(argv);
+        let mut args = macfuse::fuse_args {
+            argc: argv.len() as i32,
+            argv: argv.as_mut_ptr(),
+            allocated: 0,
+        };
+
+        let fd = macfuse::fuse_mount_compat25(
```
Comment (xujihui1985) [inline src/transport/macfuse/session.rs:256]: macfuse::fuse_mount_compat25 depend on c libfuse, maybe we can port this method into rust in the future

File: src/transport/macfuse/mod.rs [hunk 4]
```diff
@@ -0,0 +1,1177 @@
...
+use crate::BitmapSlice;
+
+mod session;
+mod macfuse;
+pub use session::*;
+use crate::transport::FileVolatileSlice;
+
+/// Error codes for Virtio queue related operations.
+#[derive(Debug)]
+pub enum Error {
```
Comment (xujihui1985) [inline src/transport/macfuse/mod.rs:27]: maybe it's better to extract this Error enum into a separate mod, that can be referenced by both macfuse and fusedev feature

**Review by bergwolf — COMMENTED** [review]
I have some comments but overall lgtm. Please add github action for it so that it is actually built and tested in CI. Thanks for the outstanding work!

File: build.rs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+#[cfg(feature = "macfuse")]
+extern crate bindgen;
+
+use std::env;
+use std::path::PathBuf;
+
+#[cfg(feature = "macfuse")]
+fn link_macfuse() {
+    println!("cargo:rustc-link-lib=fuse");
```
[the remark this answers is not in the document]
Comment (killagu) [reply build.rs:9]: For now, It's dependent `mount_darwin.c`. I can try implement it with rust.

**Comment by killagu** [comment]
Thanks for review. I'll solve the comments soon.

File: build.rs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        // Tell cargo to invalidate the built crate whenever any of the
+        // included header files changed.
+        .parse_callbacks(Box::new(bindgen::CargoCallbacks))
+        // Finish the builder and generate the bindings.
+        .generate()
+        // Unwrap the Result and panic on failure.
+        .expect("Unable to generate bindings");
+
+    // Write the bindings to the $OUT_DIR/bindings.rs file.
+    let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply build.rs:30]: It's a special env var for build.rs provided by cargo:)
https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts

File: Cargo.toml [hunk 7]
```diff
@@ -39,10 +49,11 @@ vm-memory = { version = "0.7", features = ["backend-mmap", "backend-bitmap"] }
 
 [features]
 default = ["fusedev"]
-#async-io = ["async-trait", "futures", "iou", "ringbahn"]
-fusedev = ["vmm-sys-util"]
-virtiofs = ["virtio-queue"]
-vhost-user-fs = ["virtiofs", "vhost"]
+macfuse = ["built", "bindgen", "core-foundation-sys", "diskarbitration-sys"]
```
[the remark this answers is not in the document]
Comment (killagu) [reply Cargo.toml:52]: Good point. Only one feature is better.

File: src/abi/linux_abi.rs [hunk 8]
```diff
@@ -11,13 +11,20 @@ use std::mem;
 use bitflags::bitflags;
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply src/abi/linux_abi.rs:11]: Should we split them into linux_abi and macos_abi instead? so we will have much less conditional compilations.

File: src/abi/linux_abi.rs [hunk 9]
```diff
@@ -545,27 +552,36 @@ pub struct Attr {
...
     pub uid: u32,
     pub gid: u32,
     pub rdev: u32,
+    #[cfg(target_os = "macos")]
+    pub flags: u32,
     pub blksize: u32,
+    #[cfg(target_os = "linux")]
     pub flags: u32,
+    #[cfg(target_os = "macos")]
+    pub padding: u32,
```
[the remark this answers is not in the document]
Comment (killagu) [reply src/abi/linux_abi.rs:573]: No, with the `repr(C)` order is important, It should same as macfuse definition.

File: src/abi/linux_abi.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (bergwolf) [reply src/abi/linux_abi.rs:11]: My impression is that the Macos FUSE ABI is a subset of the Linux one.

File: src/abi/linux_abi.rs [hunk 10]
```diff
@@ -576,21 +592,24 @@ impl Attr {
             atimensec: st.st_atime_nsec as u32,
             mtimensec: st.st_mtime_nsec as u32,
             ctimensec: st.st_ctime_nsec as u32,
-            mode: st.st_mode,
+            mode: st.st_mode as u32,
```
Comment (jiangliu) [inline src/abi/linux_abi.rs:595]: Should conditional compile with `#[cfg(target_os = "macos")]`.

Comment (killagu) [reply src/abi/linux_abi.rs:595]: > https://doc.rust-lang.org/reference/expressions/operator-expr.html#type-cast-expressions

> Casting between two integers of the same size (e.g. i32 -> u32) is a no-op (Rust uses 2's complement for negative values of fixed integers)

From the reference, same size cast has no cost.

File: src/api/filesystem/async_io.rs [hunk 11]
```diff
@@ -12,6 +12,13 @@ use std::time::Duration;
 
 use async_trait::async_trait;
 
+#[cfg(target_os = "linux")]
+use libc::{stat64};
+
+#[cfg(target_os = "macos")]
+use libc::{stat as stat64};
+
+
```
Comment (jiangliu) [inline src/api/filesystem/async_io.rs:21]: we could reexport stat64 from transport/abi/linux_abi, so we could avoid conditional compilation here:)

File: src/api/filesystem/mod.rs [hunk 12]
```diff
@@ -20,6 +20,18 @@ pub use fuse::OpenOptions;
...
+use libc::{
+    ino64_t,
+    stat64,
+};
+
+#[cfg(target_os = "macos")]
+use libc::{
+    ino_t as ino64_t,
+    stat as stat64,
+};
```
[the remark this answers is not in the document]
Comment (jiangliu) [reply src/api/filesystem/mod.rs:33]: How about moving this platform dependent data structure definition into linux_abi.rs?

File: src/api/server/mod.rs [hunk 13]
```diff
@@ -35,7 +35,10 @@ mod async_io;
 mod sync_io;
 
 /// Maximum buffer size of FUSE requests.
+#[cfg(target_os = "linux")]
 const MAX_BUFFER_SIZE: u32 = 1 << 20;
+#[cfg(target_os = "macos")]
+const MAX_BUFFER_SIZE: u32 = 1 << 25;
 const DIRENT_PADDING: [u8; 8] = [0; 8];
```
Comment (jiangliu) [inline src/api/server/mod.rs:42]: Please move this into linux_abi.rs too:)

File: src/lib.rs [hunk 14]
```diff
@@ -102,6 +102,8 @@ pub type Result<T> = ::std::result::Result<T, Error>;
 
 pub mod abi;
 pub mod api;
+
+#[cfg(all(feature = "fusedev", feature = "virtiofs", not(feature = "macfuse")))]
```
Comment (jiangliu) [inline src/lib.rs:106]: why passthrough doesn't work on macos?

File: src/abi/linux_abi.rs (hunk 10 above)
Comment (jiangliu) [reply src/abi/linux_abi.rs:595]: No worry, that should be zero-cost abstraction.
The compiler is very very smart now, even smarter than my expectation:)

File: src/api/server/mod.rs (hunk 13 above)
Comment (bergwolf) [reply src/api/server/mod.rs:42]: `MAX_BUFFER_SIZE` is not a kernel ABI. Instead, it is part of the server API that can be negotiated with the kernel to set the max IO buffer.

File: src/lib.rs (hunk 14 above)
Comment (killagu) [reply src/lib.rs:106]: passthrough has many syscall, It's very hard to port.

**Comment by killagu** [comment]
@bergwolf @xujihui1985 @jiangliu Please review again has refactor with helpful comments. And macos ci has added to github action, please approve to run it.

**Review by bergwolf — COMMENTED** [review]
LGTM! Just a few minor comments.

File: src/api/server/mod.rs [hunk 15]
```diff
@@ -115,6 +119,7 @@ struct ServerVersion {
 struct ServerUtil();
 
 impl ServerUtil {
+    #[allow(clippy::uninit_vec)]
```
[the remark this answers is not in the document]
Comment (killagu) [reply src/api/server/mod.rs]: It's duplicated.

File: src/api/server/sync_io.rs [hunk 16]
```diff
@@ -638,7 +641,6 @@ impl<F: FileSystem + Sync, D: AsyncDrive> Server<F, D> {
                 let mut out = InitOut {
                     major: KERNEL_VERSION,
                     minor: KERNEL_MINOR_VERSION,
-                    max_readahead,
```
[the remark this answers is not in the document]
Comment (killagu) [reply src/api/server/sync_io.rs]: My bad, it should not dropped.

File: src/transport/fusedev/macos_session.rs [hunk 17]
```diff
@@ -0,0 +1,613 @@
+// Copyright 2020-2021 Ant Group. All rights reserved.
```
[the remark this answers is not in the document]
Comment (killagu) [reply src/transport/fusedev/macos_session.rs]: It should be 2022. And other files update to 2022 too.

**Review by bergwolf — APPROVED** [review]
Thanks!

**Pull request merged**